Src/PerformanceCollector/Shared/Implementation/StandardPerformanceCollector/StandardPerformanceCollector.cs

```diff
@@ -0,0 +1,365 @@
+                        originalString,
+                        reportAs,
+                        usesInstanceNamePlaceholder,
+                        isCustomCounter,
+                        !firstReadOk,
+                        performanceCounter.CategoryName,
+                        performanceCounter.CounterName,
+                        performanceCounter.InstanceName);
+
+                string key = this.GenerateKeyForPerformanceCounter(perfData);
+                if (this.dictionary.ContainsKey(key))
+                {
+                    this.dictionary.Remove(key);
+                }
+
+                this.dictionary.Add(key, performanceCounter);
+                this.performanceCounters.Add(perfData);
+            }
+        }
+        
+        /// <summary>
+        /// Performs collection for all registered counters.
+        /// </summary>
+        /// <param name="onReadingFailure">Invoked when an individual counter fails to be read.</param>
+        public IEnumerable<Tuple<PerformanceCounterData, float>> Collect(
+            Action<string, Exception> onReadingFailure = null)
+        {
+            return this.performanceCounters.Where(pc => !pc.IsInBadState).SelectMany(
+                pc =>
+                    {
+                        float value;
+
+                        try
+                        {
+                            value = CollectCounter(this.dictionary[this.GenerateKeyForPerformanceCounter(pc)]);
+                        }
+                        catch (InvalidOperationException e)
+                        {
+                            if (onReadingFailure != null)
+                            {
+                                onReadingFailure(
+                                    PerformanceCounterUtility.FormatPerformanceCounter(this.dictionary[this.GenerateKeyForPerformanceCounter(pc)]),
+                                    e);
+                            }
+
+                            return new Tuple<PerformanceCounterData, float>[] { };
+                        }
+
+                        return new[] { Tuple.Create(pc, value) };
+                    });
+        }
+
+        /// <summary>
+        /// Creates a metric telemetry associated with the PerformanceCounterData, with the respective float value.
+        /// </summary>
+        /// <param name="pc">PerformanceCounterData for which we are generating the telemetry.</param>
+        /// <param name="value">The metric value for the respective performance counter data.</param>
+        /// <returns>Metric telemetry object associated with the specific counter.</returns>
+        public MetricTelemetry CreateTelemetry(PerformanceCounterData pc, float value)
+        {
+            var metricName = !string.IsNullOrWhiteSpace(pc.ReportAs)
+                                 ? pc.ReportAs
+                                 : string.Format(
+                                     CultureInfo.InvariantCulture,
```

Comment:
Actually yes, I somehow overlooked it. This can be moved to Module. This function is using PerformanceCounterData, which is abstracted for any collector. Will update that in the next PR/Commit!